plasmapy/plasma/grids.py

```diff
@@ -601,18 +629,100 @@ def _make_mesh(self, start, stop, num, **kwargs):
 
         return pts0, pts1, pts2
 
+    # *************************************************************************
+    # Methods
+    # *************************************************************************
+
+    def on_grid(self, pos):
+        r"""
+        Given a list of positions, determines which are in the region
+        bounded by the grid points.
+
+        For non-uniform grids, "on grid" is defined as being bounded by
+        grid points in all axes.
+
+        Parameters
+        ----------
+        pos : np.ndarray or u.Quantity array, shape (n,3)
+            An array of positions in space, where the second dimension
+            corresponds to the three dimensions of the grid.
+
+        """
+
+        if hasattr(pos, "unit"):
+            pos = pos.si.value
+
+        # Find the bounds
+        if self.is_uniform:
+            ax0_min, ax0_max = np.min(self.ax0.si.value), np.max(self.ax0.si.value)
+            ax1_min, ax1_max = np.min(self.ax1.si.value), np.max(self.ax1.si.value)
+            ax2_min, ax2_max = np.min(self.ax2.si.value), np.max(self.ax2.si.value)
+
+        else:
+            pts0, pts1, pts2 = self.grids
+            ax0_min, ax0_max = np.min(self.pts0).si.value, np.max(self.pts0).si.value
+            ax1_min, ax1_max = np.min(self.pts1).si.value, np.max(self.pts1).si.value
+            ax2_min, ax2_max = np.min(self.pts2).si.value, np.max(self.pts2).si.value
+
+        # Check each point elementwise against the bounds
+        on_grid = (
+            np.greater(ax0_min, pos[:, 0]).astype(np.int8)
+            + np.less(ax0_max, pos[:, 0]).astype(np.int8)
+            + np.greater(ax1_min, pos[:, 1]).astype(np.int8)
+            + np.less(ax1_max, pos[:, 1]).astype(np.int8)
+            + np.greater(ax2_min, pos[:, 2]).astype(np.int8)
+            + np.less(ax2_max, pos[:, 2]).astype(np.int8)
+        )
+
+        return np.where(on_grid == 0, True, False)
```

Comment:
Okay, I'm trying and failing to understand this implementation. 

Am I right in thinking that this method only actually returns indices; not a (n,3) array of float positions but a (n,) array of booleans?

Reply:
Yes, right now it takes an array of particle positions [N,3] and returns an array of Booleans [N] which indicate whether the particles are on or off the grid. 

Really I'm just interested in the total number on the grid, so if you think it would be otherwise more useful we could try to return a list of the indices of particles that are on the grid (then I can just use the length of that array)? Maybe then we could only interpolate and push those particles?

Reply:
I went with the less intrusive change and simplified the [N] array output version a little. You should, however, be able to use that array to interpolate/push only those particles, though!